src/rust/ide/src/double_representation/module.rs

```diff
@@ -2,18 +2,81 @@
 
 use crate::prelude::*;
 
+use crate::double_representation::ReferentName;
 use crate::double_representation::definition;
 use crate::double_representation::definition::DefinitionProvider;
 
 use ast::crumbs::ChildAst;
 use ast::crumbs::ModuleCrumb;
 use ast::known;
 use ast::BlockLine;
-use data::text::ByteIndex;
 use enso_protocol::language_server;
 use serde::Deserialize;
 use serde::Serialize;
-use crate::model::module::ReferentName;
+
+
+
+// ==========
+// === Id ===
+// ==========
+
+/// The segments of module name. Allow finding module in the project.
+///
+/// Example: `["Parent","Module_Name"]`
+///
+/// Includes segments of module path but *NOT* the project name (see: `QualifiedName`).
+#[derive(Clone,Debug,Shrinkwrap,PartialEq,Eq,Hash)]
+pub struct Id {
+    /// The vector is non-empty.
+    segments:Vec<ReferentName>
+}
+
+impl Id {
+    /// Construct a module's ID value from a name segments sequence.
+    ///
+    /// Fails if the given sequence is empty.
+    pub fn new(segments:impl IntoIterator<Item=ReferentName>) -> Result<Id,EmptyName> {
+        let segments = segments.into_iter().collect_vec();
+        if segments.is_empty() {
+            Err(EmptyName)
+        } else {
+            Ok(Id {segments})
+        }
+    }
+
+    /// Construct a module's ID value from a name segments sequence.
+    ///
+    /// Fails if the sequence is empty or if any of the segments is not a valid referent name.
+    pub fn try_new(segments:impl IntoIterator<Item:AsRef<str>>) -> FallibleResult<Id> {
+        let texts = segments.into_iter();
+        let names = texts.map(|text| ReferentName::new(text.as_ref()));
+
+        let segments:Vec<_> = Result::from_iter(names)?;
+        Self::new(segments).map_err(Into::into)
+    }
+
+    /// Build module's qualified name for a project with a given name.
+    pub fn qualified(&self, project_name:impl Into<String>) -> QualifiedName {
+        // TODO why safe?
+        QualifiedName::from_segments(project_name,&self.segments).unwrap()
+    }
```

Comment:
In fact it is not safe, because user can give a name what is not a valid project name.

Reply:
Good catch, method is unused, I wanted to remove it but somehow missed.